cli/command/stack/kubernetes/conversion.go

```diff
@@ -65,13 +68,44 @@ func toSwarmProtocol(protocol apiv1.Protocol) swarm.PortConfigProtocol {
 	return swarm.PortConfigProtocol("unknown")
 }
 
-func fetchPods(namespace string, pods corev1.PodInterface) ([]apiv1.Pod, error) {
-	labelSelector := labels.SelectorForStack(namespace)
-	podsList, err := pods.List(metav1.ListOptions{LabelSelector: labelSelector})
-	if err != nil {
-		return nil, err
+func fetchPods(stackName string, pods corev1.PodInterface, f filters.Args) ([]apiv1.Pod, error) {
+	services := f.Get("service")
+	nodes := f.Get("node")
+	// for existing script compatibility, support either <servicename> or <stackname>_<servicename> format
+	stackNamePrefix := stackName + "_"
+	for _, s := range services {
+		if strings.HasPrefix(s, stackNamePrefix) {
+			services = append(services, strings.TrimPrefix(s, stackNamePrefix))
+		}
+	}
+	labelSelector := labels.SelectorForStack(stackName, services...)
+	var result []apiv1.Pod
+	if len(nodes) == 0 {
+		podsList, err := pods.List(metav1.ListOptions{LabelSelector: labelSelector})
+		if err != nil {
+			return nil, err
+		}
+		result = podsList.Items
+	} else {
+		result = []apiv1.Pod{}
+		for _, n := range nodes {
+			podsList, err := pods.List(metav1.ListOptions{LabelSelector: labelSelector, FieldSelector: "spec.nodeName=" + n})
```

Comment:
This might result in tons of `pods.List()` calls. Can we not do like `spec.nodeName in (node1,node2,...)`? If not, we should probably just do one `pods.List()` call and post-filter the results (or just remove the `node` arg altogether).

Reply:
I have simplified the code and done the node filtering after querying for a full ```pods.List()```, this is much clearer in the end. Thanks for suggesting this!